cmd/gateway/zerostor/gateway-zerostor.go

```diff
@@ -0,0 +1,1080 @@
+					firstPart = false
+
+				} else {
+					offset = 0
+				}
+
+				readLength := r.Obj.Metadata.Size - offset
+				if remaingLength < readLength {
+					readLength = remaingLength
+				}
+
+				if err := zstor.Read(&r.Obj.Metadata, writer, offset, readLength); err != nil {
+					return err
+				}
+				remaingLength -= readLength
+				if remaingLength == 0 {
+					break
+				}
+			}
+			accSize += r.Obj.Size
+
+		}
+
+	}
+	return nil
+}
+
+func (zo *zerostorObjects) GetObjectInfo(ctx context.Context, bucket, object string, opts minio.ObjectOptions) (objInfo minio.ObjectInfo, err error) {
+	log.WithFields(log.Fields{
+		"bucket": bucket,
+		"object": object,
+	}).Debug("GetObjectInfo")
+
+	objInfo, err = zo.meta.GetObjectInfo(bucket, object)
+	if err != nil {
+		err = zstorToObjectErr(errors.WithStack(err), Operation("GetObjectInfo"), bucket, object)
+	}
+
+	return objInfo, err
+}
+
+func (zo *zerostorObjects) ListObjects(ctx context.Context, bucket, prefix, marker, delimiter string,
+	maxKeys int) (result minio.ListObjectsInfo, err error) {
+	log.WithFields(log.Fields{
+		"bucket":    bucket,
+		"prefix":    prefix,
+		"marker":    marker,
+		"delimiter": delimiter,
+		"maxKeys":   maxKeys,
+	}).Debug("ListObjects")
+
+	// get objects
+	result, err = zo.meta.ListObjects(ctx, bucket, prefix, marker, delimiter, maxKeys)
+	if err != nil {
+		err = zstorToObjectErr(errors.WithStack(err), Operation("ListObjects"), bucket)
+		return
+	}
+	return result, nil
+}
+
+// ListObjectsV2 implementation
+func (zo *zerostorObjects) ListObjectsV2(ctx context.Context, bucket, prefix, continuationToken, delimiter string, maxKeys int, fetchOwner bool, startAfter string) (result minio.ListObjectsV2Info, err error) {
+	log.WithFields(log.Fields{
+		"bucket":            bucket,
+		"prefix":            prefix,
+		"continuationToken": continuationToken,
+		"delimiter":         delimiter,
+		"maxKeys":           maxKeys,
+		"fetchOwner":        fetchOwner,
+		"startAfter":        startAfter,
+	}).Debug("ListObjects")
+
+	// get objects
+	result, err = zo.meta.ListObjectsV2(ctx, bucket, prefix, continuationToken, delimiter, maxKeys, fetchOwner, startAfter)
+	if err != nil {
+		err = zstorToObjectErr(errors.WithStack(err), Operation("ListObjectsV2"), bucket)
+		return
+	}
+	return result, nil
+}
+
+// PutObject implements ObjectLayer.PutObject
+func (zo *zerostorObjects) PutObject(ctx context.Context, bucket, object string, data *minio.PutObjReader, opts minio.ObjectOptions) (objInfo minio.ObjectInfo, err error) {
+	if zo.isReadOnly() {
+		err = ErrReadOnlyZeroStor
+		return
+	}
+
+	log.WithFields(log.Fields{
+		"bucket":   bucket,
+		"object":   object,
+		"metadata": opts.UserDefined,
+	}).Debug("PutObject")
+
+	return zo.putObject(ctx, bucket, object, data, opts)
+}
+
+func (zo *zerostorObjects) putObject(ctx context.Context, bucket, object string, data *minio.PutObjReader, opts minio.ObjectOptions) (objInfo minio.ObjectInfo, err error) {
```

Comment:
As discussed via telegram, this implementation is buggy and introduce a data race instead i suggest the following implementation (pseudo code)

```go

func WriterStream(cb func() (Metadata, error)) error {
	for {
		meta, err := cb()
		if err == io.EOF{
			return nil
		} else if err != nil {
			return err
		}

		//Write, and link your meta here
		//keep track if what u need to track

	}
}

func PutObject() {

	err := WriterStream(func() (Medadata, error) {
		// each call of course should read the next part of the stream 
		if noMoreData {
			return io.EOF
		}

		meta, err := zstor.Write(...)
		if err != nil {
			return err
		}
	})
}
```

Reply:
done